### PR TITLE
[inductor][NVGEMM] Make scaled-GEMM enums picklable; keep scaled autotune in-process

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #190643
* #190642
* #189841
* #189807
* #189806
* #189781
* #189780
* #189779
* #189778
* #189777
* __->__ #189776

torch.nn.functional.ScalingType/SwizzleType are pybind11 C++ enums whose class
__qualname__ is `_ScalingType`/`_SwizzleType`, but torch.nn.functional only binds
the public names `ScalingType`/`SwizzleType`. Pickle's default reduction rebuilds
the value by embedding the class object and looking it up by its private
qualname, so `getattr(torch.nn.functional, "_ScalingType")` fails -- both the
value and the class are unpicklable. This blocked shipping a scaled-GEMM
benchmark request (which carries these enums) to the autotuning subprocess pool.

Register a copyreg reducer (_register_enum_pickling, run at import) that pickles
each enum by its `.name` string and reconstructs via the public alias, so no
class object ends up in the pickle.

Pickling is necessary but not sufficient to parallelize scaled autotune: once the
pool is actually engaged (many scaled configs), the scaled precompile worker
initializes CUDA and then forks -> "Cannot re-initialize CUDA in forked
subprocess". Until that worker is made fork-safe (or the pool forced to spawn),
scaled-GEMM choices are kept on the in-process compile path via an explicit
variant guard in select_algorithm; the pickling fix is the prerequisite
groundwork for lifting that guard later.

Test Plan:
Verified the enums now round-trip through pickle after importing the utils module
(which runs the registration):

```
python -c "import pickle; import torch._inductor.codegen.nv_universal_gemm.nv_universal_gemm_utils; \
from torch.nn.functional import ScalingType, SwizzleType; \
[print(pickle.loads(pickle.dumps(v))==v) for v in (ScalingType.BlockWise1x16, ScalingType.BlockWise1x32, SwizzleType.SWIZZLE_32_4_4, SwizzleType.NO_SWIZZLE)]"
```

Confirmed the remaining fork-CUDA blocker by removing the guard and running a
scaled NVFP4 autotune with the full config set (~190 choices): the PicklingError
is gone (the request reaches the worker) and it now surfaces the "Cannot
re-initialize CUDA in forked subprocess" error instead, which is why the guard is
retained.

Authored with assistance from Claude Code.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by mlazos** [comment]
## AI-Review

> *Automated multi-agent review (2 Claude + 2 Codex + synthesis). This is AI-generated — please verify findings manually.*

---

## Critical Issues (must fix)

- **Enum pickling fix lives in the wrong module and runs as an import side effect.** `nv_universal_gemm_utils.py` (`_register_enum_pickling`, ~lines 12-41) mutates process-wide `copyreg` state at import time to repair pickling of `torch.nn.functional.ScalingType`/`SwizzleType`. Picklability is a contract of those public pybind types, not of an Inductor NVGEMM utility. As written, the fix only applies if this backend module happens to be imported — any other path that pickles these enums (multiprocessing, caching, other serialization) silently gets the broken default. Move the reducer to the enum definitions (fix the pybind `__qualname__`/module) or register it from the owning `torch.nn.functional` module, and prefer an explicit idempotent init over a bare import side effect. *(Consensus: Claude Design #1/#2, Codex Design #1 — 3 reviewers.)*

- **Generic scheduler is coupled to a backend-specific variant string.** `select_algorithm.py:~4779` gates on `c.bmreq.variant.name != "SCALED_GEMM"`, hardcoding one backend's fork-safety limitation into `AlgorithmSelectorCache` and silently overriding the configured subprocess policy. Model subprocess compatibility as an explicit capability on the benchmark request/backend (e.g. `bmreq.supports_subprocess`) so the selector picks execution mode without knowing individual variants. Otherwise this branch grows into an unmaintainable allowlist. *(Consensus: Claude Design #5, Codex Design #2 — 2 reviewers.)*

## Important Suggestions (should fix)

- **New cross-process contracts lack regression coverage.** Add tests for enum serialization round-trip in a fresh process (without depending on NVGEMM import order) and for execution-mode selection via the proposed capability contract. *(Codex Design #3.)*

- **Fork-safety is a systemic issue, not a per-variant stopgap.** The root cause — the precompile worker initializes CUDA and the pool then forks — will resurface for any variant that touches CUDA before fork. File a tracked issue for the durable fix (spawn context, or defer CUDA init past fork) rather than leaving an inline TODO. *(Claude Design #5, first bullet.)*

- **The pickling fix may now be dead relative to its stated purpose.** With `SCALED_GEMM` forced in-process, no remaining variant ships `ScalingType`/`SwizzleType` requests to the subprocess pool, so `_register_enum_pickling` no longer serves the rationale in its comment. Reconcile this: either the fork-safety exclusion or the pickling workaround is redundant as scoped. Resolving the two critical items above (upstream pickling + capability flag) likely clarifies which is actually needed. *(Claude Bug, non-bug observation.)*

## Minor Notes (consider)

- **Fragile enum-class acquisition.** `type(ScalingType.BlockWise1x16)` / `type(SwizzleType.NO_SWIZZLE)` derive the class from a specific member; renaming/removing that member breaks module import, not just the feature. Reference the class directly. (Claude Bug confirmed it won't raise today, so this is hardening, not a bug.) *(Claude Design #3.)*

- **Consolidate scattered enum defenses.** The `.name` keying in `_operand_sig` and the `.name` pickling reducer are separate symptoms of "external C++ enums are poorly-behaved Python objects." Rather than patching each call site as it breaks, consider a single documented adapter that normalizes external enums into stable, hashable, picklable forms at the subsystem boundary. *(Claude Design #4.)*

---

**Note on bug reviews:** Both bug reviewers found no functional defects — the mechanisms (`.name` keying, reduce-by-string, `and` short-circuit fallthrough) are verified correct. All blocking feedback is about placement, ownership, and API design, not correctness. Claude Bug notes its pickle round-trip was verified by reading code/CPython semantics, not by execution.

**Review by karthickai — APPROVED** [review]

**Review by crcrpar — APPROVED** [review]
